### ISSUE TITLE
GetApplicableArchivePaths ignores ModKey parameter

### ISSUE THREAD

**Issue by focustense** [Mutagen-Modding/Mutagen#207]
I can see in the [source](https://github.com/Mutagen-Modding/Mutagen/blob/2a9362cfb219f04ab41e31cb872c07d871218867/Mutagen.Bethesda.Core/Archives/Archive.cs#L90) that it actually isn't used at all.

This is no particular problem for me, since I was previously using XEditLib's `GetLoadedContainers` which also had no filtering, and had already designed things to work around that kind of structure. But at first I assumed that the key _was_ used, which led me to first enumerate all the loaded mods and then enumerate the BSAs for "each", leading to massive duplication in the list and some subsequent exceptions when trying to load them into a dictionary. Right now I've switched to passing `ModKey.Null` and only doing the one call, which works fine.

This parameter should either be respected, or removed.

**Comment by Noggog** [comment]
Archive applicable paths got some work done it it in the dev branch
https://github.com/Mutagen-Modding/Mutagen/blob/dev/Mutagen.Bethesda.Core/Archives/Archive.cs

Can import/use it by swapping to the prerelease nuget 0.30-prereleaseXYZ

Originally the parameter was exposed (but not used) because I imagined eventually ModKey would play a part in whether BSAs were applicable, I just didn't have the logic prepped in my head for what that would look like at the time

The prerelease has some more filled out concepts, and uses ModKey a bit now.  I imagine there's still a few improvements to be made in its logic, so feel free to put it through the ringer and let me know if something can be adjusted

**Comment by Noggog** [comment]
Oh, reading this again, just wanted to follow up.  I think `GetApplicableArchivePaths` is meant to be queried in relation to a single mod.  Eg, i want to load strings for Skyrim.esm, give me the BSAs I think might have them.  

You can loop multiple mods, and union all results of `GetApplicableArchivePaths` to find all the BSAs related to an active load order.. at this point I'd suggest sticking a `.Distinct()` call on there.  I'm not 100% certain on whether BSAs only apply to one mod.  I'd have to research a lot more to confirm that in order to guarantee uniqueness across a loop.

**Comment by focustense** [comment]
> You can loop multiple mods, and union all results of GetApplicableArchivePaths to find all the BSAs related to an active load order

That's what I was originally doing, with a single `SelectMany`. But because it ignored the mod key, I was getting the same BSAs repeated about 500 times. For now I'll just stick with `Null` until the dev work is pushed to release, at which point it'll probably throw an exception and I'll replace it with the old `SelectMany` loop. No big deal, package updates are never completely trouble-free. `Distinct` would be a lot slower by comparison, running on a list of 50-100k items, so I prefer not.

> I'm not 100% certain on whether BSAs only apply to one mod

It depends on what you mean by "mod". There can be multiple mods in Mod Organizer that provide the same BSA, which is something this app actually checks for. But you probably mean "mod" as ESM/ESP/ESL. Those only match archives with exactly the same name, or the same name with " - Textures" appended. The only way to have a BSA theoretically belong to multiple plugins is for someone to foolishly name their plugin as "foo - Textures.esp" (or esm/esl), which I seriously hope no one out there is doing, although nothing surprises me anymore.

Mods can also _depend on_ BSAs they don't actually own, but that's a different sort of problem. The DLCs do exactly this; Dawnguard, Dragonborn, etc. do not have archives, they just assume that what they need is in the main Skyrim archives. This is also a terrible practice that I hope no modders are doing.

I checked your code in dev a few hours ago and it looked basically correct, although IIRC accepting any hyphenated suffix might be a little too forgiving. As far as I know it's literally just "- Textures" that's special, and although it's not totally ubiquitous, plenty of plugins _do_ have hyphens in their name, so you do risk accidentally assigning the same archive to two plugins that way. Like maybe someone out there released "Vigilant - MyAwesomePatch.esp" with accompanying BSA, you might count that as a Vigilant archive which would be incorrect.

That's if I read the code right. I think that's what it's doing, just dropping anything after the last hyphen and checking if it matches the plugin.

**Comment by Noggog** [comment]
> accepting any hyphenated suffix might be a little too forgiving

Good call.  Got a unit test prepped up that shows indeed, it's too permissive.  Do we have a list of the known BSA suffixes?  
There's the ones in Skyrim.ini, like `Textures`, `Shaders`, `Misc`, etc.  But not 100% sure if the ones present there are the full range of allowed options

> Mods can also depend on BSAs they don't actually own, but that's a different sort of problem.    The DLCs do exactly this; Dawnguard, Dragonborn, etc. do not have archives, they just assume that what they need is in the main Skyrim archives.

Yeah, that's what drove this line:
https://github.com/Mutagen-Modding/Mutagen/blob/567f69e873fa29d9f80c799a3136a4aa04fe2a73/Mutagen.Bethesda.Core/Archives/Archive.cs#L130

Any bsa that is in the ini applies to anything querying for applicability.   I thought this ini was where people listed any bsas that could apply to multiple plugins that didn't match by name.   Eg, the pattern would be:
- Exact name matches
- Exact name matches, minus any known suffixes like `- Textures`
- Any bsa listed in the ini

Maybe we could dial that back to only applying to the base esms or something.   Haven't done the research.  As I've been hinting at, most of this logic is just a baseline that I'm sure needs refining after some research

**Comment by focustense** [comment]
As far as I know, "Textures" is the _only_ valid special suffix. Skyrim itself gets to have "Meshes0", "Meshes1", "Textures[0-8]" and so on - but try to do that with one of your own plugins and they'll just be ignored. That's why optimizers like CAO still have to create dummy ESL plugins for the extra archives when a mod can't fit into just one.

Maybe there is some way to manipulate which archives are loaded by the game independently of the plugins - IIRC Mod Organizer had an experimental feature like that several years ago but the devs terminated it with prejudice, possibly because it didn't work correctly or reliably. But in any case, that situation would have to be very rare, and there'd be no way to detect it from the filenames alone anyway.

If you're doing this just to be able to associate all the fancy Skyrim BSAs with the base game, and I assume similar things for other games, then I'd probably just use a literal whitelist of BSA names and never mind the prefix/suffix thing. Or if you really hate that idea, then read the `[Archive]` section from the game's INI (e.g. `sResourceArchiveList`) which will tell you exactly which ones are being loaded and therefore belong to the base game.
